### DISCUSSION TITLE
Is the book still in active development?

### DISCUSSION THREAD

**Discussion by AdamJel** [tnballo/high-assurance-rust#35] [General]
Hello,

firstly, let me thank all authors of this perfect book. It is very well structured, one of the best ratio of information/density.

That being said, I would like to ask, if the book is still in active development and there is a plan to finish it @tnballo? Last commit is 7 months old and I havent found any source of news like blog or twitter.

Some kind of promotion might be benefial, together with activating the option of sponsoring here on GitHub.

Thank you
Adam

**Comment by AdamJel** [comment]
Update:

I tried to contact the main author Tiemoko Ballo:
  - @tnballo 
  - [website](https://tiemoko.com)
  - [twitter](https://twitter.com/tnballo)
  - [LinkedIn](https://www.linkedin.com/in/tiemokoballo/)

but without any success.

According to his LinkedIn, he started at a new job at 1Password ~11 months ago, which might be the cause of stalled work on the project.

I don't mean to stalk him, but I am a bit frustrated I wasn't able to contact him in any way. I hope he is ok and not ill or something. If somebody wants to try to contact him, perhaps with higher influence on Twitter or if you know somebody from within 1Password or something, please, update us.

I really hope, he will re-continue effort on this project, because it is one of the best I came upon on the internet so far.

---

Aside, in the meanwhile, if you want to try some other high quality source, I can recommend Casey Muratori's material on [Performance-Aware Programming](https://www.computerenhance.com/p/table-of-contents), introduced with this [video](https://www.youtube.com/watch?v=tD5NrevFtbU)

**Comment by tnballo** [comment]
Hey Adam! We're so glad to hear you’ve found the book valuable, thanks for reaching out. Sorry for the late reply.

The halt in content production is on me, not on my awesome editors. For transparency: before the book’s launch I had the privilege of working on it full-time. It was quite an adventure. Definitely a risk, but probably the single most rewarding and energizing project I’ve ever gotten to work on.

However, shortly after the book went public I made the difficult decision to return to a full-time engineering role. Both because I missed building software with a team and because of the obvious practical benefits of traditional employment (health insurance in the US, etc).

I love my current role (get to work on world-class security products that even my non-technical friends use daily!), but to be brutally honest it’s been a challenge to grow into a new set of responsibilities while simultaneously finding the energy to power through nights and weekends of technical writing. I haven’t yet found a balance or a system that works for me and my family!

My employer is totally supportive of me pursuing my own projects on my own time, the issue is me personally not having the energy to effectively sustain 1.5-2 jobs, at this particular point in my life, without dropping output quality. I am actively trying to find a schedule and process that can make this a sustainable reality.

With that context, to answer your questions:

1. The book is under inconsistent development within a private repo (to not have to iterate/edit in the open).
2. Finishing the book is a major life goal for me. But I can’t commit to a firm timeline. Given the scope and what I know about the process behind other technical books, multi-year effort is a prerequisite. This is a long-haul game where +/- 6 months to a 1 year may be par for the course (though consistency should be the goal).

I appreciate your kind suggestion regarding promotion and sponsorship. To address each:

- Haven’t given promotion much thought in terms of a concrete strategy, but am dedicated to keeping all current and future content freely accessible online - so that anyone who enjoys the current 150+ pages of content can share it with friends freely (e.g. “word of mouth”). Whenever new chapters are ready, I hope to submit a link to the “This Week in Rust” newsletter to reach a population most likely to be interested.

- Enabling sponsorship via GitHub sponsors is a possibility I’ll have to give more thought in the future, in terms of weighing pros and cons. The book currently has more of a B2B support model, but the current roadblock is definitely time constraints for content production. I don’t know that there’s a realistic path to any B2B/B2C model becoming a viable replacement for traditional employment.

It pains me greatly that we can’t deliver the rest of the book on a consistent schedule. Especially with so many exciting advancements at the intersection of software assurance and Rust programming, I truly wish I could be more immersed and productive on this front. Being separated from writing for a while has caused me some anxiety, which is a factor my inability to stay on top of open-source maintenance tasks for this repo.

Seeing that you care about this project means a lot to me. I also care about it greatly. I invested a significant chunk of my life savings into getting it started full-time when I could. Now I need to figure out a way to continue it without burning out!

Hope this [long-winded] answer is helpful, even if it’s not the most satisfactory :)

**Reply by AdamJel** [reply]
Hi,

and thank you @tnballo for such and wide-explaining answer, it is very much appreciated.

Your time-constrain issues are absolutely understandable. And it is reassuring to hear, that you do intent to continue the effort on this project.


I do not want to impose anything onto you, as it is by far not my place, but I would like to propose something anyway, hoping you will forgive if it's too much of out place.

You stated that the main problem is in time domain, not financial one. But you also stated, that you got employed because of employment-tied benefits such as insurance. You also stated that not having enough time to work on the project causes you frustration. Also, you want to keep the project free as in free-beer.

It seems to me, that it could be possible, that if the project generated enough money for you, you could be willing to allocate yourself in full-time manner to it. IMO this is the truth for many aside-time projects - they are started with good intention, but collide with real life boundaries such as providing for your family. It is also my understanding, that even though the community (generally speaking) favorites open-source and free as in free-speech projects, it has no problem what so ever to financially support these efforts. There is quite a plethora of good will founded projects. All those kickstarters and substacks and whatnot being a proof of that.

I would like to keep my argument general, but I will link my link in above comment to Casey Muratori anyway. As it appears to me, he wanted to commit himself to performance targeted programming and tasted the water with some free content on youtube and elsewhere. It proved the existence of interest and subsequently he started a (semi?)-paid content on substack.

I am not suggesting, that you should place your content behind a paid-wall, but rather that the will of a community to pay for valuable resources is not negligible. Perhaps it would be possible to find a way how to monetize this project without sacrificing the book being free. You could e.g. do a Q&A sessions, or book content demonstrations as a paid content alongside sponsorship here on GitHub. Or something completely different. Perhaps this book being finished is of great interest to [Ferrocene](https://ferrous-systems.com/ferrocene/), maybe some cooperation could go from this direction? Or even a cooperation with Casey on some next course - mission critical as in high-assurance and high-performance? To me, that would be a match made in heaven.

My point is, that I am confident that with the right amount of reach to audience, you could allocate yourself exclusively to this project (at least for the phase of development). I myself would be perfectly willing (financially) supporting you in such a decision.



But anyway, as I stated in the begging, it is not my place to impose anything onto you. Rather I selfishly wish for this book to continue.

However this ends up, I wish you to find the perfect balance for you and your family and overall well being.

Thank you for creating this massive effort in the first place. I cannot overstate how appreciated it is, hopefully speaking on behalf of wider community.

**Comment by davidr6432** [comment]
Hi,

I just wanted to say that I don’t mind paying for high quality content e.g. eBook.

For example, Jon Gjengset wrote “Rust for Rustaceans” (https://nostarch.com/rust-rustaceans).

Thanks,

From: Tiemoko Ballo ***@***.***>
Sent: Wednesday, April 26, 2023 11:59 PM
To: tnballo/high-assurance-rust ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [EXTERNAL] Re: [tnballo/high-assurance-rust] Is the book still in active development? (Discussion #35)


Hey Adam! We're so glad to hear you’ve found the book valuable, thanks for reaching out. Sorry for the late reply.

The halt in content production is on me, not on my awesome editors. For transparency: before the book’s launch I had the privilege of working on it full-time. It was quite an adventure. Definitely a risk, but probably the single most rewarding and energizing project I’ve ever gotten to work on.

However, shortly after the book went public I made the difficult decision to return to a full-time engineering role. Both because I missed building software with a team and because of the obvious practical benefits of traditional employment (health insurance in the US, etc).

I love my current role (get to work on world-class security products that even my non-technical friends use daily!), but to be brutally honest it’s been a challenge to grow into a new set of responsibilities while simultaneously finding the energy to power through nights and weekends of technical writing. I haven’t yet found a balance or a system that works for me and my family!

My employer is totally supportive of me pursuing my own projects on my own time, the issue is me personally not having the energy to effectively sustain 1.5-2 jobs, at this particular point in my life, without dropping output quality. I am actively trying to find a schedule and process that can make this a sustainable reality.

With that context, to answer your questions:

  1.  The book is under inconsistent development within a private repo (to not have to iterate/edit in the open).
  2.  Finishing the book is a major life goal for me. But I can’t commit to a firm timeline. Given the scope and what I know about the process behind other technical books, multi-year effort is a prerequisite. This is a long-haul game where +/- 6 months to a 1 year may be par for the course (though consistency should be the goal).

I appreciate your kind suggestion regarding promotion and sponsorship. To address each:

  *   Haven’t given promotion much thought in terms of a concrete strategy, but am dedicated to keeping all current and future content freely accessible online - so that anyone who enjoys the current 150+ pages of content can share it with friends freely (e.g. “word of mouth”). Whenever new chapters are ready, I hope to submit a link to the “This Week in Rust” newsletter to reach a population most likely to be interested.
  *   Enabling sponsorship via GitHub sponsors is a possibility I’ll have to give more thought in the future, in terms of weighing pros and cons. The book currently has more of a B2B support model, but the current roadblock is definitely time constraints for content production. I don’t know that there’s a realistic path to any B2B/B2C model becoming a viable replacement for traditional employment.

It pains me greatly that we can’t deliver the rest of the book on a consistent schedule. Especially with so many exciting advancements at the intersection of software assurance and Rust programming, I truly wish I could be more immersed and productive on this front. Being separated from writing for a while has caused me some anxiety, which is a factor my inability to stay on top of open-source maintenance tasks for this repo.

Seeing that you care about this project means a lot to me. I also care about it greatly. I invested a significant chunk of my life savings into getting it started full-time when I could. Now I need to figure out a way to continue it without burning out!

Hope this [long-winded] answer is helpful, even if it’s not the most satisfactory :)

—
Reply to this email directly, view it on GitHub<https://github.com/tnballo/high-assurance-rust/discussions/35#discussioncomment-5732072>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AXSJCEBQCMBAJF3SKW5WLCDXDES2FANCNFSM6AAAAAAXAFRJSY>.
You are receiving this because you are subscribed to this thread.Message ID: ***@***.******@***.***>>


________________________________

IMPORTANT CONFIDENTIALITY NOTICE:

This e-mail (including any documents referred to in, or attached to, this e-mail) may contain information that is personal, confidential or the subject of copyright, privilege or other proprietary rights in favour of Aristocrat, its affiliates (which includes Pixel United, Big Fish Games, and Product Madness) or third parties. This e-mail is intended only for the named addressee. Any privacy, confidentiality, legal professional privilege, copyright or other proprietary rights in favour of Aristocrat, its affiliates (which includes Pixel United, Big Fish Games, and Product Madness) or third parties, is not lost nor waived if this e-mail has been sent to you in error.

If you have received this e-mail in error you should: (i) not copy, disclose, distribute or otherwise use it or its contents without the consent of Aristocrat or the owner of the relevant rights; and (ii) delete it from your system and destroy all copies.

Any personal information contained in this e-mail must be handled in accordance with applicable privacy laws.

Electronic and internet communications can be interfered with or affected by viruses and other defects. As a result, such communications may not be successfully received or, if received, may cause interference with the integrity of receiving, processing or related systems (including hardware, software and data or information on, or using, that hardware or software). Aristocrat gives no assurances and accepts no liability in relation to these matters.

**Comment by AdamJel** [comment]
* REPOST for inactivity
* ANY update for fans at all, please?

---

Hi,

and thank you @tnballo for such and wide-explaining answer, it is very much appreciated.

Your time-constrain issues are absolutely understandable. And it is reassuring to hear, that you do intent to continue the effort on this project.

I do not want to impose anything onto you, as it is by far not my place, but I would like to propose something anyway, hoping you will forgive if it's too much of out place.

You stated that the main problem is in time domain, not financial one. But you also stated, that you got employed because of employment-tied benefits such as insurance. You also stated that not having enough time to work on the project causes you frustration. Also, you want to keep the project free as in free-beer.

It seems to me, that it could be possible, that if the project generated enough money for you, you could be willing to allocate yourself in full-time manner to it. IMO this is the truth for many aside-time projects - they are started with good intention, but collide with real life boundaries such as providing for your family. It is also my understanding, that even though the community (generally speaking) favorites open-source and free as in free-speech projects, it has no problem what so ever to financially support these efforts. There is quite a plethora of good will founded projects. All those kickstarters and substacks and whatnot being a proof of that.

I would like to keep my argument general, but I will link my link in above comment to Casey Muratori anyway. As it appears to me, he wanted to commit himself to performance targeted programming and tasted the water with some free content on youtube and elsewhere. It proved the existence of interest and subsequently he started a (semi?)-paid content on substack.

I am not suggesting, that you should place your content behind a paid-wall, but rather that the will of a community to pay for valuable resources is not negligible. Perhaps it would be possible to find a way how to monetize this project without sacrificing the book being free. You could e.g. do a Q&A sessions, or book content demonstrations as a paid content alongside sponsorship here on GitHub. Or something completely different. Perhaps this book being finished is of great interest to [Ferrocene](https://ferrous-systems.com/ferrocene/), maybe some cooperation could go from this direction? Or even a cooperation with Casey on some next course - mission critical as in high-assurance and high-performance? To me, that would be a match made in heaven.

My point is, that I am confident that with the right amount of reach to audience, you could allocate yourself exclusively to this project (at least for the phase of development). I myself would be perfectly willing (financially) supporting you in such a decision.

But anyway, as I stated in the begging, it is not my place to impose anything onto you. Rather I selfishly wish for this book to continue.

However this ends up, I wish you to find the perfect balance for you and your family and overall well being.

Thank you for creating this massive effort in the first place. I cannot overstate how appreciated it is, hopefully speaking on behalf of wider community.

**Reply by tnballo** [reply]
Hey Adam,

Thanks for your feedback and consideration! 

Avenues to monetization are something I’ll have to give careful consideration in the future. Right now, my full-time job allows me to grow as a security-focused software developer (my background is in security research). Working on a product that protects the secrets of millions of users is an incredibly valuable experience. In the long (multi-year) term, I’m confident this experience and growth will allow me to write a better book that’s more useful to a wide range of developers.

I've had a chance to refine a chunk of the WIP content into an update - see #37 (v4.0.0-alpha). This is not a massive release, about 13 pages + diagrams. But my hope is that, as is the goal for the book overall, it offers unique technical insights in an accessible form. 

For me personally, releasing even a small update is motivating and helps build momentum for ironing out the next chapter (in terms of what needs to be cut and added to make it viable).

Though not currently seeking collaborations, integrating Ferrocene work and recent news is definitely on the roadmap for book updates! Specifically:

* An upcoming section of chapter 4 looks at DO-332 in the context of [this public paper](https://www.adacore.com/uploads/techPapers/DynamicMemoryManagement.pdf) by AdaCore and Ferrous Systems.

    * Aim to increase alignment between the book and industry standards, with an eye toward generally applicable patterns.

* Chapter 3 content regarding a Rust language specification will be updated to reflect both [Ferrocene’s specification work](https://ferrous-systems.com/blog/the-ferrocene-language-specification-is-here/) and [MiniRust](https://github.com/RalfJung/minirust).

    * Specification efforts are incredibly valuable to the broader community and pave the way for Rust’s adoption in safety-critical settings.

Cheers,
Tiemoko
